working/macros/feature-specification.md

```diff
@@ -736,41 +736,59 @@ to introspect over non-macro metadata annotations applied to declarations.
+All `MetadataAnnotation` objects have a `Code get code` getter, which gives
+access to the annotation as a `Code` object.
+
+In addition, there will be two subtypes of `MetadataAnnotation`:
+
+- `IdentifierMetadataAnnotation`: A simple const identifier, has a single
+  `Identifier get identifier` getter.
+- `ConstructorMetadataAnnotation`: A const constructor invocation. This will
+  have the following getters:
+  - `Identifier get type`
+  - `Identifier get constructor`
+  - `Arguments get arguments`
+    - The `Arguments` class will provide access to the positional and named
+      arguments as separate `Code` objects.
 
-Since annotations may contain references to types or identifiers that the macro
-does not import, we choose to expose a more abstract API (similar to
-[DartObject][]).
+For any macro which only wants to emit code from annotations back into the
+program, these `Code` objects are sufficient.
 
-**TODO**: Define the exact API.
+For a macro which wants to access the actual _value_ of a given argument or
+the metadata annotation as a whole, they can evaluate `Code` instances as
+constants (see next section).
 
-#### Annotations that require macro expansion
+### Constant evaluation
 
-This could happen if the annotation class has macros applied to it, or if
-some argument(s) to the annotation constructor use macros.
+Macros may want the ability to evaluate constant expressions, in particular
+those found as arguments to metadata annotations.
 
-Because macros are not allowed to generate code that shadows an identifier
-in the same library, we know that if an annotation class or any arguments to it
-could be resolved, then we can assume that resolution is correct.
+We expose this ability through the `DartObject evaluate(Code code)` api, which
+is available in all phases, with the following restrictions:
 
-This allows us to provide an API for macro authors to attempt to evaluate an
-annotation in _any phase_. The API may fail (if it requires more macro
-expansion to be done), but that is not expected to be a common situation. In
-the case where it does fail, users should typically be able to move some of
-their code to a separate library (which they import). Then things from that
-library can safely be used in annotations in the current library, and evaluated
-by macros.
+- No identifier in `code` may refer to a constant which refers to any
+  system environment variable, Dart define, or other configuration which is not
+  otherwise visible to macros.
+- All identifiers in `code` must be defined outside of current strongly
+  connected component (that is, the strongly connected component which triggered
+  the current macro expansion).
 
-Evaluation must fail if there are any macros left to be expanded on the
-annotation class or any arguments to the annotation constructor.
+The `DartObject` api is an abstract representation of an object, which can
+represent types which are not visible to the macro itself. It will closely
```

Comment:
I mean types which do not appear in the transitive imports of the macro library (or even the entire program created for running the macro).

Reply:
Ah, now I understand, thank you.
Well, this is probably true for most types that macros get through introspection, right?
But I can see that it might be necessary to remind it, why we cannot get actual object, only some view.

Reply:
👍